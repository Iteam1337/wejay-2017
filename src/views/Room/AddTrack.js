import React, { Component } from 'react'
import { roomQuery } from './Room'
import { gql, graphql } from 'react-apollo'
import md5 from 'md5'

type AddTrackProps = {
  mutate: Function,
  roomName: string
}

export class AddTrack extends Component {
  props: AddTrackProps

  handleKeyUp = evt => {
    const { mutate, roomName } = this.props

    if (evt.keyCode === 13) {
      const userId = md5(localStorage.getItem('user'))

      mutate({
        variables: {
          input: {
            roomName,
            userId,
            spotifyId: evt.target.value
          }
        },
        optimisticResponse: {
          queueTrack: {
            album: [],
            artists: [],
            duration: 0,
            name: '',
            spotifyUri: evt.target.value,
            user: '',
            __typename: 'Track'
          }
        },
        update: (store, { data: { queueTrack } }) => {
          const data = store.readQuery({
            query: roomQuery,
            variables: {
              name: roomName
            }
          })

          const exists = data.room.queue.find(
            track => track.spotifyUri === queueTrack.spotifyUri
          )

          const notCurrent =
            data.room.currentTrack &&
            data.room.currentTrack.spotifyUri !== queueTrack.spotifyUri

          if (!exists && notCurrent) {
            data.room.queue.push(queueTrack)
          }

          store.writeQuery({
            query: roomQuery,
            variables: {
              name: roomName
            },
            data
          })
        }
      })

      evt.target.value = ''
    }
  }

  render () {
    return (
      <div>
        <input
          onKeyUp={this.handleKeyUp}
          placeholder="Spotify URI"
          type="text"
        />
      </div>
    )
  }
}

const addTrackMutation = gql`
  mutation queueTrack($input: QueueInput!) {
    queueTrack(input: $input) {
      album {
        images {
          url
        }
        name
      }
      artists {
        name
      }
      duration
      name
      spotifyUri
      user
    }
  }
`

export default graphql(addTrackMutation)(AddTrack)
