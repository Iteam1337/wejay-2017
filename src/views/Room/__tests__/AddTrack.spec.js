import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AddTrack } from '../AddTrack'
import md5 from 'md5'

jest.mock('md5', () => jest.fn(() => 'md5'))
jest.mock('../Room', () => ({
  roomQuery: 'roomQuery'
}))

describe('components/AddTrack', () => {
  let component
  let mutate
  let instance

  beforeEach(() => {
    mutate = jest.fn()

    component = shallow(<AddTrack mutate={mutate} roomName="test" />)

    instance = component.instance()
  })

  it('can render', () => {
    expect(toJson(component)).toMatchSnapshot()
  })

  describe('#handleKeyUp', () => {
    it('does nothing if keypress is not enter (13)', () => {
      const event = {
        keyCode: 31
      }

      instance.handleKeyUp(event)

      expect(mutate).not.toBeCalled()
    })

    it('fetches the saved user and hashes the email', () => {
      const event = {
        keyCode: 13,
        target: {
          value: '1234'
        }
      }

      localStorage.getItem.mockReturnValue('cookie@monster.com')

      instance.handleKeyUp(event)

      expect(localStorage.getItem).toHaveBeenCalledWith('user')
      expect(md5).toHaveBeenCalledWith('cookie@monster.com')
    })

    it('calls mutation with variables', () => {
      const event = {
        keyCode: 13,
        target: {
          value: '1234'
        }
      }

      instance.handleKeyUp(event)

      expect(mutate.mock.calls[0][0].variables).toMatchSnapshot()
    })

    it('sets optimistic response', () => {
      const event = {
        keyCode: 13,
        target: {
          value: '1234'
        }
      }

      instance.handleKeyUp(event)

      expect(mutate.mock.calls[0][0].optimisticResponse).toMatchSnapshot()
    })

    it('handles updates', () => {
      const event = {
        keyCode: 13,
        target: {
          value: '1234'
        }
      }

      instance.handleKeyUp(event)

      const update = mutate.mock.calls[0][0].update
      const store = {
        readQuery: jest.fn().mockReturnValue({
          room: {
            currentTrack: {
              spotifyUri: '4321'
            },
            queue: []
          }
        }),
        writeQuery: jest.fn()
      }
      const updateData = {
        data: {
          queueTrack: {
            spotifyUri: '1234'
          }
        }
      }

      update(store, updateData)

      expect(store.readQuery).toHaveBeenCalledWith({
        query: 'roomQuery',
        variables: {
          name: 'test'
        }
      })

      expect(store.writeQuery).toHaveBeenCalledWith({
        query: 'roomQuery',
        variables: {
          name: 'test'
        },
        data: store.readQuery()
      })

      expect(event.target.value).toEqual('')
    })
  })
})
