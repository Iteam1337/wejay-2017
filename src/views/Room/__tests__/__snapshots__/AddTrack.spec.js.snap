// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddTrack #handleKeyUp calls mutation with variables 1`] = `
Object {
  "input": Object {
    "roomName": "test",
    "spotifyId": "1234",
    "userId": "md5",
  },
}
`;

exports[`components/AddTrack #handleKeyUp sets optimistic response 1`] = `
Object {
  "queueTrack": Object {
    "__typename": "Track",
    "album": Array [],
    "artists": Array [],
    "duration": 0,
    "name": "",
    "spotifyUri": "1234",
    "user": "",
  },
}
`;

exports[`components/AddTrack can render 1`] = `
<div>
  <input
    onKeyUp={[Function]}
    placeholder="Spotify URI"
    type="text"
  />
</div>
`;
